Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+#### Scheduling background work
+
+In May [AWS described in a blog post that you can run background tasks in Lambda](https://aws.amazon.com/blogs/compute/running-code-after-returning-a-response-from-an-aws-lambda-function/) until the runtime asks for more work from the control plane. We want to support this by adding new API that allows background processing, even after the response has been returned.
+
+## Proposed Solution
+
+#### async/await-first API
+
+Large parts of `Lambda`, `LambdaHandler`, and `LambdaRuntime` will be re-written to use async/await constructs in place of the `EventLoop` family of interfaces.
+
+#### Providing ownership of main() and support for swift-service-lifecycle
+
+* Instead of conforming to a handler protocol, users can now create a `LambdaRuntime` by passing in a handler closure.
+* `LambdaRuntime` conforms to `ServiceLifecycle.Service` by implementing a `run()` method that contains initialization and graceful shutdown logic.
+* This allows the lifecycle of the `LambdaRuntime` to be managed with `swift-service-lifecycle` *alongside* and in the same way the lifecycles of the required services are managed, e.g. `try await ServiceGroup(services: [postgresClient, ..., lambdaRuntime], ...).run()`.
+* Dependencies can now be injected into `LambdaRuntime` — `swift-service-lifecycle` guarantees that the services will be initialized *before* the `LambdaRuntime`’s `run()` function is called.
+* The required services can then be used within the handler in a structured concurrency manner. `swift-service-lifecycle` takes care of listening for termination signals and terminating the services as well as the `LambdaRuntime` in correct order.
+* `LambdaTerminator` can now be eliminated because its role is replaced with `swift-service-lifecycle`. The termination logic of the Lambda function will be implemented in the conforming `run()` function of `LambdaRuntime`.
+
+With this, the earlier code snippet can be replaced with something much easier to read, maintain, and debug:
+
+```swift
+/// Instantiate services
+let postgresClient = PostgresClient()
+
+/// Instantiate LambdaRuntime with a closure handler implementing the business logic of the Lambda function
+let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+    /// Use initialized service within the handler
+    try await postgresClient.query(...)
+}
+
+/// Use ServiceLifecycle to manage the initialization and termination
+/// of the services as well as the LambdaRuntime
+let serviceGroup = ServiceGroup(
+    services: [postgresClient, runtime],
+    configuration: .init(gracefulShutdownSignals: [.sigterm]),
+    logger: logger
+)
+try await serviceGroup.run()
+```
+
+#### Simplifying Codable support
+
+A detailed explanation is provided in the Codable Support section. In short, much of the boilerplate code defined for each handler protocol in `Lambda+Codable` and `Lambda+String` will be replaced with a single `LambdaCodableAdapter` class.
```

Comment:
Why is the adapter a `class`?

Reply:
Hahaha. Leftover has become a struct in our prototype already.